### ISSUE TITLE
Slightly cryptic error message for missing index.rst file.

### ISSUE THREAD

**Issue by jskksj** [sphinx-doc/sphinx#2197]
I had deleted the current index.rst file and not yet replaced it while running 'make html'.  The error message requested a issue report should this be a user error.

```
# Sphinx version: 1.3.3
# Python version: 3.4.3 (CPython)
# Docutils version: 0.12 release
# Jinja2 version: 2.8
# Last messages:
#   none found
#   pickling environment...
#   done
#   checking consistency...
#   /home/jsk/anaconda3/envs/cv/src/calibration/doc/CameraCalibration.rst:: WARNING: document isn't included in any toctree
#   done
#   preparing documents...
#   done
#   writing output... [ 50%] CameraCalibration
#   writing output... [100%] index
# Loaded extensions:
#   alabaster (0.7.7) from /home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/alabaster/__init__.py
Traceback (most recent call last):
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/cmdline.py", line 240, in main
    app.build(opts.force_all, filenames)
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/application.py", line 266, in build
    self.builder.build_update()
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/builders/__init__.py", line 252, in build_update
    'out of date' % len(to_build))
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/builders/__init__.py", line 323, in build
    self.write(docnames, list(updated_docnames), method)
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/builders/__init__.py", line 361, in write
    self._write_serial(sorted(docnames), warnings)
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/builders/__init__.py", line 369, in _write_serial
    self.write_doc(docname, doctree)
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/builders/html.py", line 444, in write_doc
    ctx = self.get_doc_context(docname, body, metatags)
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/builders/html.py", line 412, in get_doc_context
    self_toc = self.env.get_toc_for(docname, self)
  File "/home/jsk/anaconda3/envs/cv/lib/python3.4/site-packages/sphinx/environment.py", line 1223, in get_toc_for
    tocdepth = self.metadata[docname].get('tocdepth', 0)
KeyError: 'index'
```

**Comment by tk0miya** [comment]
This is came from all builders 
https://github.com/sphinx-doc/sphinx/blob/master/sphinx/builders/__init__.py#L346

**Comment by tk0miya** [comment]
In my short investigation:
This is came from the behavior that Sphinx always rewrites `master_doc` even if it is not found.
https://github.com/sphinx-doc/sphinx/blob/master/sphinx/builders/__init__.py#L346 (in `write()`)

This error happens on incremental build.
At the same time, I got another error on full build:

```
# Sphinx version: 1.3.3+
# Python version: 2.7.11 (CPython)
# Docutils version: 0.12 release
# Jinja2 version: 2.8
# Last messages:
#   none found
#   pickling environment...
#   done
#   checking consistency...
#   /Users/tkomiya/work/tmp/foo/foo.md:: WARNING: document isn't included in any toctree
#   /Users/tkomiya/work/tmp/foo/projects/index.rst:: WARNING: document isn't included in any toctree
#   done
#   preparing documents...
#   done
#   writing output... [ 25%] foo
# Loaded extensions:
#   alabaster (0.7.7) from /Users/tkomiya/work/tmp/testbed/lib/python2.7/site-packages/alabaster-0.7.7-py2.7.egg/alabaster/__init__.pyc
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/cmdline.py", line 244, in main
    app.build(opts.force_all, filenames)
  File "/Users/tkomiya/work/sphinx/sphinx/application.py", line 266, in build
    self.builder.build_update()
  File "/Users/tkomiya/work/sphinx/sphinx/builders/__init__.py", line 251, in build_update
    'out of date' % len(to_build))
  File "/Users/tkomiya/work/sphinx/sphinx/builders/__init__.py", line 322, in build
    self.write(docnames, list(updated_docnames), method)
  File "/Users/tkomiya/work/sphinx/sphinx/builders/__init__.py", line 360, in write
    self._write_serial(sorted(docnames), warnings)
  File "/Users/tkomiya/work/sphinx/sphinx/builders/__init__.py", line 368, in _write_serial
    self.write_doc(docname, doctree)
  File "/Users/tkomiya/work/sphinx/sphinx/builders/html.py", line 445, in write_doc
    self.handle_page(docname, ctx, event_arg=doctree)
  File "/Users/tkomiya/work/sphinx/sphinx/builders/html.py", line 792, in handle_page
    output = self.templates.render(templatename, ctx)
  File "/Users/tkomiya/work/sphinx/sphinx/jinja2glue.py", line 138, in render
    return self.environment.get_template(template).render(context)
  File "/Users/tkomiya/work/tmp/testbed/lib/python2.7/site-packages/Jinja2-2.8-py2.7.egg/jinja2/environment.py", line 989, in render
    return self.environment.handle_exception(exc_info, True)
  File "/Users/tkomiya/work/tmp/testbed/lib/python2.7/site-packages/Jinja2-2.8-py2.7.egg/jinja2/environment.py", line 754, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/tkomiya/work/sphinx/sphinx/themes/basic/page.html", line 10, in top-level template code
    {%- extends "layout.html" %}
  File "/Users/tkomiya/work/sphinx/sphinx/themes/classic/layout.html", line 13, in top-level template code
    {% set script_files = script_files + ['_static/sidebar.js'] %}
  File "/Users/tkomiya/work/sphinx/sphinx/themes/basic/layout.html", line 160, in top-level template code
    {%- block content %}
  File "/Users/tkomiya/work/sphinx/sphinx/themes/basic/layout.html", line 178, in block "content"
    {%- block sidebar2 %}{{ sidebar() }}{% endblock %}
  File "/Users/tkomiya/work/sphinx/sphinx/themes/basic/layout.html", line 178, in block "sidebar2"
    {%- block sidebar2 %}{{ sidebar() }}{% endblock %}
  File "/Users/tkomiya/work/tmp/testbed/lib/python2.7/site-packages/Jinja2-2.8-py2.7.egg/jinja2/sandbox.py", line 355, in call
    return __context.call(__obj, *args, **kwargs)
  File "/Users/tkomiya/work/sphinx/sphinx/themes/basic/layout.html", line 66, in template
    {%- block sidebartoc %}
  File "/Users/tkomiya/work/sphinx/sphinx/themes/basic/layout.html", line 67, in block "sidebartoc"
    {%- include "localtoc.html" %}
  File "/Users/tkomiya/work/tmp/foo/_templates/localtoc.html", line 2, in top-level template code
    {% set toctree = toctree(maxdepth=2, collapse=False, includehidden=True) %}
  File "/Users/tkomiya/work/tmp/testbed/lib/python2.7/site-packages/Jinja2-2.8-py2.7.egg/jinja2/sandbox.py", line 355, in call
    return __context.call(__obj, *args, **kwargs)
  File "/Users/tkomiya/work/sphinx/sphinx/builders/html.py", line 782, in <lambda>
    ctx['toctree'] = lambda **kw: self._get_local_toctree(pagename, **kw)
  File "/Users/tkomiya/work/sphinx/sphinx/builders/html.py", line 718, in _get_local_toctree
    docname, self, collapse, **kwds))['fragment']
  File "/Users/tkomiya/work/sphinx/sphinx/environment.py", line 1238, in get_toctree_for
    doctree = self.get_doctree(self.config.master_doc)
  File "/Users/tkomiya/work/sphinx/sphinx/environment.py", line 1272, in get_doctree
    f = open(doctree_filename, 'rb')
IOError: [Errno 2] No such file or directory: u'/Users/tkomiya/work/tmp/foo/_build/doctrees/index.doctree'
```

It is caused from this line:
https://github.com/sphinx-doc/sphinx/blob/master/sphinx/environment.py#L1238 (in `get_toctree_for()`)
There is more function referencing the `master_doc`.

Many features of Sphinx depend on existence of `master_doc`.
So it should raise error on if `master_doc` not found (instead of warning).

**Comment by tk0miya** [comment]
@birkenfeld Is there any reason to only warn if master_doc not found?
This behavior was implemented at very early Sphinx (before v0.2!).
https://github.com/sphinx-doc/sphinx/commit/50da49fab419949106442e5eec47831d0c82d0b8#diff-4f9fdea88baff0b430508f9793ae9c93R358

If possible, I would raise error in that case (before users got any exceptions).
Because many features of Sphinx depends on existence of master_doc.
Or, we'll have to handle these exceptions for docs without master_doc.
(but I could not imagine that case)

**Comment by birkenfeld** [comment]
I think you're right, this should be an error (especially since it leads to exceptions like this.)

**Comment by tk0miya** [comment]
I understand. I'll fix it later :-)

**Comment by tk0miya** [comment]
FIxed at 7bb685b.
Thank you for reporting!
